core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -203,6 +208,11 @@ class LogManager(val logDirs: Array[File],
                          delay = InitialTaskDelayMs,
                          period = flushCheckpointMs,
                          TimeUnit.MILLISECONDS)
+      scheduler.schedule("kafka-delete-logs",
```

Comment:
The scheduler uses `ScheduledThreadPoolExecutor`, which has a fixed number of threads and an unbounded queue. Using this scheduler for IO risks blocking all the threads in the thread-pool if/when disk IO takes too long or never returns. As this thread-pool executor is used for other periodic tasks such as, log-retention, log-flushing, and recovery-checkpointing, we definitely don't want to lock-up all the threads doing nothing. Direct use of  `scheduler.schedule` opens up this unlikely but dangerous possibility. This is perhaps strong enough reason to stay away from direct scheduling and use an explicit queue of logs such that only one thread from the pool is used at a time.